archive/unarchive.go

```diff
@@ -91,6 +92,49 @@ func Untar(src io.Reader, dst string, opts ...UnarchiveOption) error {
 
 	tr := tar.NewReader(src)
 
+	if uc.stripIfOnlyDir {
+		// Duplicate src so that we can reset it.
+		var buf bytes.Buffer
+		if _, err := io.Copy(&buf, src); err != nil {
+			return fmt.Errorf("could not copy reader: %v", err)
+		}
+
+		// Reset the buffer on exit
+		defer buf.Reset()
+
+		// Set the tarball readers by using the previously copied buffer.
+		tr = tar.NewReader(bytes.NewReader(buf.Bytes()))
+		tr2 := tar.NewReader(bytes.NewReader(buf.Bytes()))
+
+		// Map to track top-level directory entries
+		topLevelDirs := make(map[string]struct{})
+
+		for {
+			header, err := tr2.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return fmt.Errorf("failed to read tar header: %w", err)
+			}
+
+			// Split path components to identify the top-level directory
+			components := strings.SplitN(header.Name, "/", 2)
+			if len(components) > 0 && components[0] != "pax_global_header" {
+				topLevelDirs[components[0]] = struct{}{}
+			}
+
+			// If more than one top-level directory is detected, exit
+			if len(topLevelDirs) > 1 {
+				break
+			}
+		}
+
+		if len(topLevelDirs) == 1 {
+			uc.stripComponents = 1
+		}
+	}
+
```

Comment:
Question: isn't it faster to just unpack as is and then `mv` everything one level down? The operation itself does not extra copying compared to here (I can see at least 1 copy)

Reply:
It's at least more memory efficient to do it this way since there's no intermediate buffer, but i'm not sure whether it's necessarily faster since you anyway have to iterate over at least the first-level child entries and move those to the new location.